### PR TITLE
[notebook] Support multiple docker images for notebook

### PR DESCRIPTION
**Pull request by tpoterba**
Preparation for adding ccg-workshop materials

### CODE REVIEW

File: env-setup.sh [hunk 1]
```diff
@@ -65,7 +65,7 @@ then
 fi
 
 kubectl version -c || gcloud components install kubectl
-kubectl version || gcloud container clusters get-credentials $CLUSTER_NAME
+gcloud container clusters get-credentials $CLUSTER_NAME --region us-central1-a
```
Comment (danking) [inline env-setup.sh]: restore `kubectl version || `

File: notebook/Makefile [hunk 2]
```diff
@@ -1,32 +1,43 @@
...
+build/images/%/_PUSHED: build/images/%/_BUILT
+	docker tag $* $(IMAGE)
+	docker push $(IMAGE)
+	echo $(IMAGE) > $@
+
+
+notebook-worker-images: $(PUSHED)
+	cat $^ > $@
+
+
```
Comment (danking) [inline notebook/Makefile]: one newline between each rule

File: notebook/Makefile [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
+build: $(BUILT) notebook-worker-images
+	echo $(BUILT)
```
Comment (danking) [inline notebook/Makefile]: delete

File: notebook/Makefile [hunk 4]
```diff
@@ -1,32 +1,43 @@
...
 	docker build . -t notebook
 
+
```
Comment (danking) [inline notebook/Makefile]: delete extra newline

File: notebook/notebook/notebook.py [hunk 5]
```diff
@@ -56,15 +57,17 @@ def read_string(f):
 log.info(f'INSTANCE_ID {INSTANCE_ID}')
 
 try:
-    with open('notebook-worker-image', 'r') as f:
-        WORKER_IMAGE = f.read().strip()
+    with open('notebook-worker-images', 'r') as f:
+        def get_name(line):
+            return re.search("/[^/:]+:", line).group()[1:-1]
```
Comment (danking) [inline notebook/notebook/notebook.py]: What does `group` do with no arguments? It doesn't look like that regex has any groups? This is what I would've expected: `re.search("/([^/:]+):", line).group(1)`

File: notebook/notebook/templates/index.html [hunk 6]
```diff
@@ -10,9 +10,17 @@
       <h1>Hail Notebooks</h1>
       <form action="{{ form_action_url }}", method="post">
         <p>
-          <label for="password">Class Key:</label>
+          <label>Class Key:</label>
```
Comment (danking) [inline notebook/notebook/templates/index.html]: I think the `for` link should stay? It still describes the input box named password.

File: notebook/Makefile [hunk 7]
```diff
@@ -1,32 +1,43 @@
-.PHONY: build push run-docker run deploy clean-workers build-worker push-worker
+.PHONY: build push run-docker run deploy
```
Comment (danking) [inline notebook/Makefile]: can you add a phony `clean` target that `rm -rf build`?

File: notebook/Makefile [hunk 8]
```diff
@@ -1,32 +1,43 @@
...
 
 PROJECT = $(shell gcloud config get-value project)
+BUILT = $(shell ls -1 images | awk '{print "build/images/" $$0 "/_BUILT"}')
+PUSHED = $(shell ls -1 images | awk '{print "build/images/" $$0 "/_PUSHED"}')
 
-build: push-worker
+build/images/%/_BUILT: images/%/Dockerfile $(shell find images/$*)
```
Comment (danking) [inline notebook/Makefile]: You forgot to check in these docker files.

Comment (tpoterba) [reply notebook/Makefile]: oops

File: notebook/notebook/notebook.py (hunk 5 above)
Comment (tpoterba) [reply notebook/notebook/notebook.py]: dunno, this worked though. Your way is better

File: notebook/notebook/templates/index.html (hunk 6 above)
Comment (tpoterba) [reply notebook/notebook/templates/index.html]: I got unhappy highlighting in IntelliJ. I put it back

**Comment by tpoterba** [comment]
not testing?

**Comment by tpoterba** [comment]
ah, that would make sense. It seemed like it was taking too long...

**Pull request merged**